stripe_test.go

```diff
@@ -118,6 +119,71 @@ func TestDo_Retry(t *testing.T) {
 	assert.Equal(t, 2, requestNum)
 }
 
+// Tests client retries when HTTP returns timeout.
+//
+// To run only this test:
+//
+//     go test . -run TestDo_RetryOnTimeout -test.v
+//
+func TestDo_RetryOnTimeout(t *testing.T) {
+	type testServerResponse struct {
+		Message string `json:"message"`
+	}
+
+	timeout := time.Second
+	requestNum := 0
+
+	testServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		err := r.ParseForm()
+		assert.NoError(t, err)
+
+		// The body should always be the same with every retry. We've
+		// previously had regressions in this behavior as we switched to HTTP/2
+		// and `Request` became non-reusable, so we want to check it with every
+		// request.
+		assert.Equal(t, "bar", r.Form.Get("foo"))
+		time.Sleep(timeout)
+		requestNum++
```

Comment:
Okay, this is a little subtle, but there's a race here which is causing the Travis build to fail on this test. Check there for details, but you should also be able to reproduce it locally by using `go test` with the `-race` flag (which is what we do in Travis):

```
go test . -run TestDo_RetryOnTimeout -test.v -race
```

I believe what's happening is that you have multiple server Goroutines that are active in this handler concurrently. Because the client is timing out early, the first request is still active on the server when the second is handled. The fix is pretty easy though: make sure to use a mutex around any variable accesses like this one that could be touched from multiple Goroutines.